The Great Embooleanating

timwoj opened the pull request:
This PR makes the massive switch from using `int` types for boolean values to using `bool` types for most of them. This has bugged me for a while and I thought I'd finally take care of it.

There's a few other things in here:
- Some of the `Dispatch` methods for timers where not marked as `override`. These have been fixed.
- The logic in `Event::Describe` changed a little bit, but the old logic didn't make a lot of sense to begin with.
- `NVT_Analyzer`'s constructor changes from calling the default initializer on basic types to using in-class intiialization.
- I left the bit-set-being-used-as-booleans in `Conn` alone, since we already know changing that causes performance issues from a previous PR.
- Removed the `virtual` keyword from `NCP_Session::Deliver`, since that class is never inherited from anywhere.
- Removed unnecessary `const` from a number of method arguments in `SMTP_Analyzer`

I tried to make as few changes to virtual methods that may be overridden so as to minimize API breakage. I admittedly didn't go through all of the call sites for the methods I changed so I'm sure I missed places that could be converted as well, but this is the large majority of it.

timwoj commented:
One of the big tests is failing on every platform. I'll take a look.

timwoj commented:
Tests are fixed. A branch with the same name needs to be merged on the `zeek-testing-private` repo.

jsiwek commented:
> Tests are fixed. A branch with the same name needs to be merged on the `zeek-testing-private` repo.

I didn't look closely at the commits yet, but wouldn't expect those baselines to change for this type of refactor.  Some even look more than just potentially-superficial connection reordering and have flipped the `orig`/`resp` roles.  Did you find an explanation for those that confirms the new output is actually "more correct" ?

timwoj commented:
Fixed the issue with DNS and reverted the change for the private testing repo. FreeBSD is still failing on CI because of it just randomly giving up in the middle of testing sometimes.

timwoj commented:
Added a few more commits after I remembered that clang-tidy has the modernize-use-bool-literal checker, and ran it through that.

rsmmr commented on src/analyzer/protocol/tcp/TCP.cc:
```diff
@@ -478,8 +478,8 @@ static void syn_weirds(TCP_Flags flags, TCP_Endpoint* endpoint, int data_len)
-			int len, int is_orig, TCP_Flags flags,
-			int& do_close, int& gen_event)
+			int len, bool is_orig, TCP_Flags flags,
+			bool& do_close, bool& gen_event)
```
Note for future work: non-const pass-by-reference would be good to replace with pass-by-pointer.

timwoj commented:
Is it even possible to mark a virtual method as deprecated and have the compiler complain about someone overriding it? The compiler would warn if someone tried to call the not-overridden version of the method but if the method is overridden it's technically a new method and not deprecated. For example, the following snippet doesn't output any warnings:

```
class A {
public:
	[[deprecated("use bool")]]
	virtual void a(int val) {}
	};

class B : public A
	{
public:
	void a(int val) override {}
	};

int main(int argc, char** argv)
	{
	B b;
	b.a(1);
	}
```

It also doesn't output warnings if the version in `A` is pure virtual.

rsmmr commented:
Yeah, I think you're right, I was hoping "deprecated" would lead to a warning when overridding, but I also don't manage to trigger that. Then the only other approach I would see is keeping the old version in the base class as well, and calling it from the new version to trigger any legacy overrides that may still be around. However, I'm not really proposing that: it'd mean a 2nd virtual dispatch for each call, and worse, it also may lead to messiness handling return values correctly.

Seems the real solution is for people to use `override` concisely.

Ok, so I'm fine leaving these as is unless anybody has a better idea still.

rsmmr commented:
So I think all this needs is a rebase.

timwoj commented:
@rsmmr Rebase is done and I added a NEWS entry about the API changes
